core/matchers/src/test/groovy/au/com/dius/pact/core/matchers/JsonBodyMatcherSpec.groovy

```diff
@@ -328,4 +329,315 @@ class JsonBodyMatcherSpec extends Specification {
+
+    where:
+    actual                                                    | matches
+    '[{"id":"b", "status":"up"},{"id":"a", "status":"down"}]' | false
+    '[{"id":"a", "status":"up"},{"id":"b", "status":"down"}]' | true
+    '[{"id":"a", "status":"up"},{"id":"b", "status":"up"}]'   | true
+    '[{"id":"b", "status":"down"},{"id":"a", "status":"up"}]' | true
+  }
+
+  @Unroll
+  def 'matching json bodies - with ignore-order - and mixed elements'() {
+    given:
+    matchers.addCategory('body')
+            .addRule('$', IgnoreOrderMatcher.INSTANCE)
+            .addRule('$[*].id', new EqualsMatcher())
+            .addRule('$[0].status', new EqualsMatcher())
+            .addRule('$[1].status', new RegexMatcher('up|down'))
+            .addRule('$[3]', TypeMatcher.INSTANCE)
+    def actualBody = OptionalBody.body(actual.bytes)
+    def expectedBody = OptionalBody.body('[{"id":"a", "status":"up"}, 4, {"id":"b", "status":"down"}]'.bytes)
+
+    expect:
+    matcher.matchBody(expectedBody, actualBody, true, matchers).empty == matches
+
+    where:
+    actual                                                        | matches
+    '[{"id":"a", "status":"up"}, 4, {"id":"b", "status":"down"}]' | true
+    '[{"id":"b", "status":"up"}, {"id":"a", "status":"up"}, 5]'   | true
+    '[{"id":"b", "status":"down"}, {"id":"a", "status":"up"}, 5]' | true
+    '[{"id":"b", "status":"up"}, {"id":"a", "status":"down"}, 5]' | false
+    '[{"id":"c", "status":"up"}, {"id":"a", "status":"up"}, 5]'   | false
+  }
+
+  @Unroll
+  def 'matching json bodies - with ignore-order - and elements without unique ids'() {
+    given:
+    matchers.addCategory('body')
+            .addRule('$', IgnoreOrderMatcher.INSTANCE)
+            .addRule('$[0].id', new RegexMatcher('a|b'))
+            .addRule('$[0].status', new EqualsMatcher())
+            .addRule('$[1].id', new RegexMatcher('b|c'))
+            .addRule('$[1].status', new RegexMatcher('up|down'))
+    def expectedBody = OptionalBody.body('[{"id":"a", "status":"up"},{"id":"b", "status":"down"}]'.bytes)
+    def actualBody = OptionalBody.body(actual.bytes)
+
+    expect:
+    matcher.matchBody(expectedBody, actualBody, true, matchers).empty == matches
+
+    where:
+    actual                                                    | matches
+    '[{"id":"b", "status":"up"},{"id":"b", "status":"down"}]' | true  // expct[0]==actl[0] & expct[1]==actl[0|1]
+    '[{"id":"a", "status":"up"},{"id":"b", "status":"down"}]' | true  // expct[0]==actl[0] & expct[1]==actl[1]
+    '[{"id":"a", "status":"up"},{"id":"b", "status":"up"}]'   | true  // expct[0]==actl[0] & expct[1]==actl[1]
+    '[{"id":"b", "status":"down"},{"id":"c", "status":"up"}]' | false // expct[0] no match & expct[1]==actl[1]
+// I don't feel this test should fail - let's discuss if you disagree
+//    '[{"id":"b", "status":"up"},{"id":"c", "status":"up"}]'   | false // expct[0|1]==actl[0], no unique for all expct
```

Comment:
I must not understand what you mean by 'no unique for all expected'

Expected [{"id":"a","status":"up"}, {"id":"b","status":"down"}]
Actual [{“id":"b", "status":"up"},{"id":"c", "status":"up"}]

$[0].id is RegexMatcher( a | b ) - check
$[0].status is EqualsMatcher( ) - check
$[1].id is RegexMatcher( b | c ) - check
$[1].status is RegexMatcher( up | down ) - check

Reply:
Though expct[0] == actl[0] and expct[1] == actl[0], there is not a unique match for both expct[0] and expct[1].

This is just a more complicated case of expected=[100,100] and actual=[100,200].  Here, we also have expct[0] == actl[0] and expct[1] == actl[0], but because there are not two unique instances of 100 in actual we don't have a match.

Reply:
Sorry, but I must still be missing something.  To me it looks like expct[0] == actl[0] and expct[1] == actl[1] given all the matching rules.  

I added the test for  expected=[100,100] and actual=[100,200] with just an IgnoreOrderMatcher (not MinEqualsIgnoreOrderMatcher) and that passed too.  Which I would've expected.

Reply:
😅 oops, I was paying attention to the comment instead of the actual test case there...  You are :100: right, expct[0] == actl[0] and expct[1] == actl[1] for that test case (just as you have it for [L604](https://github.com/mitre/pact-jvm/blob/a57de008f7a54d7d4ea78ca39d9da763854a6a39/core/matchers/src/test/groovy/au/com/dius/pact/core/matchers/JsonBodyMatcherSpec.groovy#L604) immediately below this comment.

Given that, I'd suggest changing the test case to this - which looks like it is currently passing.

```suggestion
    '[{"id":"b", "status":"up"},{"id":"a", "status":"down"}]' | false // expct[0]==expct[1]==actl[0], no unique for each
```

Reply:
No problem.  Just want to make sure we are on the same page.  Better to go back and forth internally then find failures after we send out a merge request.